core/src/blocktree.rs

```diff
@@ -386,47 +412,145 @@ impl Blocktree {
+                });
+                (set_index..set_index + erasure_meta.config.num_coding() as u64).for_each(|i| {
+                    if index.coding().is_present(i) {
+                        if let Some(shred) = prev_inserted_codes.remove(&(slot, i)).or_else(|| {
+                            let some_code = code_cf
+                                .get_bytes((slot, i))
+                                .expect("Database failure, could not fetch code shred");
+                            if let Some(code) = some_code {
+                                bincode::deserialize(&code).ok()
+                            } else {
+                                warn!("Code shred deleted while reading for recovery");
+                                None
+                            }
+                        }) {
+                            available_shreds.push(shred);
+                        }
+                    }
+                });
+                if let Ok(mut result) = Shredder::try_recovery(
+                    &available_shreds,
+                    erasure_meta.config.num_data(),
+                    erasure_meta.config.num_coding(),
+                    set_index as usize,
+                    slot,
+                ) {
+                    recovered_data_shreds.append(&mut result.recovered_data);
+                }
+            }
+        }
+        recovered_data_shreds
+    }
+
+    pub fn insert_shreds(&self, shreds: Vec<Shred>) -> Result<()> {
         let db = &*self.db;
         let mut batch_processor = self.batch_processor.write().unwrap();
         let mut write_batch = batch_processor.batch()?;
 
-        let mut just_inserted_data_indexes = HashMap::new();
+        let mut just_inserted_data_shreds = HashMap::new();
+        let mut just_inserted_coding_shreds = HashMap::new();
+        let mut erasure_metas = HashMap::new();
         let mut slot_meta_working_set = HashMap::new();
         let mut index_working_set = HashMap::new();
 
-        shreds.iter().for_each(|shred| {
+        shreds.into_iter().for_each(|shred| {
             let slot = shred.slot();
+            let shred_index = u64::from(shred.index());
 
-            let _ = index_working_set.entry(slot).or_insert_with(|| {
+            let index_meta = index_working_set.entry(slot).or_insert_with(|| {
                 self.index_cf
                     .get(slot)
                     .unwrap()
                     .unwrap_or_else(|| Index::new(slot))
             });
-        });
-
-        // Possibly do erasure recovery here
 
-        let dummy_data = vec![];
+            if let Shred::Coding(coding_shred) = &shred {
+                // This gives the index of first coding shred in this FEC block
+                // So, all coding shreds in a given FEC block will have the same set index
+                let pos = u64::from(coding_shred.header.position);
+                if shred_index >= pos {
+                    let set_index = shred_index - pos;
```

Comment:
Referring to the comment above, if `pos` is the same, but `shred_index` is not the same for all coding_shreds in an FEC block, how will `set_index = shred_index - pos` be the same for all coding blobs in an FEC block?